New Resource: github_organization_project

shihanng opened the pull request:
Hi, this PR adds the ability to manage organization projects to this provider as described in https://developer.github.com/v3/projects/#create-an-organization-project. This is my first time working with Terraform provider, please let me know if I miss anything. Thank you.

shihanng commented:
It seems that there is conflict. Would you like me to rebase?

radeksimko commented:
@shihanng yes, I merged a few PRs lately and I'm going through others, including this one. I'm going to review it regardless, but rebasing would be incredibly helpful.

Thanks!

radeksimko commented on github/resource_github_organization_project.go:
```diff
@@ -0,0 +1,116 @@
+			"url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
```
How do you feel about coverage of the remaining fields which are available through `PATCH`?
https://developer.github.com/v3/projects/#update-a-project
specifically `state`, `organization_permission` and `public`?

shihanng replied:
:thinking: It seems that the library that we are using does now support these fields yet?
- There is a way to set the `state` via [`ProjectOptions`](https://godoc.org/github.com/google/go-github/github#ProjectOptions) but the value of the `state` is not included in [`Project`](https://godoc.org/github.com/google/go-github/github#Project) so we can't get the current state of the project via the library.
- For `organization_permission`, is seems that there is an issue opened and being worked on currently https://github.com/google/go-github/issues/884
- For `public`, is it not available via the go-github library yet.

radeksimko replied:
Ah, understood. Thanks for the detailed answer & research. We shall leave it out for the initial implementation then.

radeksimko commented on github/resource_github_organization_project.go:
```diff
@@ -0,0 +1,116 @@
+	options := github.ProjectOptions{
+		Name: n,
+		Body: b,
+	}
```
Nitpick (non-blocking): I think we could just inline the fields above to the struct to reduce LOC and improve readability a bit.

shihanng replied:
Fixed in f1a04f4, thank you.

radeksimko commented on github/resource_github_organization_project.go:
```diff
@@ -0,0 +1,116 @@
+	options := github.ProjectOptions{
+		Name: n,
+		Body: b,
+	}
```
Nitpick - as above.

shihanng replied:
Fixed in f1a04f4, thank you.

radeksimko commented on website/docs/r/organization_project.html.markdown:
```diff
@@ -0,0 +1,34 @@
+
+* `name` - The name of the project.
+
+* `body` - The body of the project.
```
Can you please document whether each field is `Optional` or `Required` here?

shihanng replied:
Added in 2ea7426. Thanks!

radeksimko commented on github/resource_github_organization_project_test.go:
```diff
@@ -0,0 +1,114 @@
+		},
+	})
+}
+
```
Do you mind adding a test for import also?

shihanng replied:
Added test for import in ec08678. Thank you.

radeksimko commented on github/resource_github_organization_project.go:
```diff
@@ -0,0 +1,116 @@
+
+	d.Set("name", project.GetName())
+	d.Set("body", project.GetBody())
+	d.Set("url", fmt.Sprintf("https://github.com/orgs/%s/projects/%d", o, project.GetNumber()))
```
Is there any reason we should be constructing the URL here instead of reading it from the API? It appears that [`html_url`](https://developer.github.com/v3/projects/#create-an-organization-project) is the attribute we could use?

shihanng replied:
Ah yes, but it is not made available in the go-github library? See [Project](https://godoc.org/github.com/google/go-github/github#Project).

shihanng commented:
@radeksimko, thank you for the reviews. I've addressed some comments in code, and also the rest as comments. Could you please take another look. Thank you.

shihanng commented:
Thank you so much 😄! I will work on #115 later on and ping you again when it is ready.
